### ISSUE TITLE
Massive dataset + data_generator

### ISSUE THREAD

**Issue by AntreasAntoniou** [maxpumperla/elephas#15]
Hello guys,

Before I begin, I want to thank you for this amazing tool. It's truely awesome and enables true distributed deep learning out of the box. Now for some questions.

1. From what I've seen I need to load the dataset in memory before I send it to elephas for distributed processing. When the dataset is massive, as in multiple times my ram, how can I use an hdf5 so that each worker can load parts of the dataset from disk and do the processing? Please let me know.

2. Is there any support for keras "data_generator" functionality? That enables real time data augmentation? 

Thank you for your support.

**Comment by maxpumperla** [comment]
Hey @AntreasAntoniou, thanks for the positive feedback, that's very motivating. You're welcome!

1. Well, you load data into an RDD (or data frame), which is distributed already. Most of the examples we provide in the example folder create an RDD from a pair of numpy arrays via a little helper function from elephas. The thing is, though, you could as well read your RDD some file system (HDFS, S3 etc.) directly. For a toy example of how to do this, check the data frame generation in this notebook:

https://github.com/maxpumperla/elephas/blob/master/examples/Spark_ML_Pipeline.ipynb

2. I haven't used data_generator yet, so I'd have to come back to you. Initial thought: real time augmentation is not exactly in the spirit of Spark data structures. Maybe Spark Streaming could be a mid-term extension.

**Comment by AntreasAntoniou** [comment]
So for data augmentation you suggest I do that locally in the hdf5 file? and create epochs in the hdf5 or something  right?

**Comment by maxpumperla** [comment]
Generally speaking, I'd prepare small data in-memory, medium-sized via some sort of streaming/augmentaion and large (don't say the magic word here) data with a suitable large-scale processing engine like Spark. For instance, to train on imagenet, we put the whole thing on S3 and use Spark's capability to load directories, as e.g. explained here:

http://stackoverflow.com/questions/27477730/spark-how-to-download-multiple-gzipped-files-from-s3-as-a-single-rdd

In case you want to work with hd5 files, maybe this one will help:

https://hdfgroup.org/wp/2015/03/from-hdf5-datasets-to-apache-spark-rdds/

Does that answer your questions?

**Comment by bhack** [comment]
@maxpumperla When you are using a model with keras Graph Api you call the fit like 
```model.fit(train={'input1':X1, 'input2':X2, 'output1':Y1, 'conv24_output':Y2})```
How could be handled with elephas helper utils with ML pipelines and dataframes?

**Comment by maxpumperla** [comment]
@bhack the same way as for ```Sequential```, elephas doesn't care which model, as long as it has ```model_from_yaml``` and ```to_yaml``` applicable, so you can borrow from this example:

https://github.com/maxpumperla/elephas/blob/master/examples/Spark_ML_Pipeline.ipynb

**Comment by bhack** [comment]
@maxpumperla Seems to me that all the fit used in Elephas are only with a single X and Y. See https://github.com/maxpumperla/elephas/blob/ea9b323b326943fd19d601b86549a84c2ff0b192/elephas/spark_model.py#L215

**Comment by maxpumperla** [comment]
@bhack you're right, I'm not accounting for multiple inputs/outputs. What's happening internally doesn't matter, though. I'm not sure how to account for this properly, but it would be nice to have. I'll have to think about this first.

**Comment by maxpumperla** [comment]
If you want, you could open another issue for that, not sure that's still the same thing.

**Comment by maxpumperla** [comment]
I think I'll wait for the release of Keras 1.0 and use the momentum to tackle this.

**Comment by EderSantana** [comment]
Hey Max, I just noticed you putting up some really good work in here. This is awesome!!!

Let me hop in this question and ask you a few things. If you believe I should start another issue, let me know! But I'd like to hear your opinion.

For several applications (at least those I'm working on :D) the data handling needs more work and the model training is a good old single GPU or at least single machine model. I wanted to build a data server that exposed a numpy-array-like interface. Something like:
```python
Data = DataServer("url", prefered_batch_size, indices, callback)
x_batch = Data[:128]
```

Then, every time we call a slice, in parallel behind the scenes the model would prepare the next batch doing all the preprocessing specified by the callback in a list of possible elements addressed by indices. The user could nevertheless get any element `Data[i]`, but not as fast as if she follow the promised order and batch-size, which is what we will be caching in memory. This would allow us to get Keras to train large datasets without having to change things in Keras. 

Also, Keras' `model.fit_generator` is interesting, but it is a whole new API. I wish we could keep one unifying fit API for keras and actually get the dataset part to be simpler.

Let me know your opinion. I was thinking should I go with Spark for that? Should I build a zeromq server? What would you suggest?

**Comment by bhack** [comment]
Ok i've opened a new issue at https://github.com/maxpumperla/elephas/issues/16

**Comment by maxpumperla** [comment]
Hey Eder, thanks man! I just remembered I still have an issue open on Agnez. Sorry, I have to get back to you soon. 

I know exactly what you mean, training is rarely the bottleneck given enough powerful GPUs, but the preprocessing can hurt. If you ask me, Spark is a pretty good choice for the latter, especially since it's built for this domain (which e.g. zeromq is not). Refering to the above ipynb again, you could for example build a preprocessing pipeline like this: 

```
from pyspark.ml.feature import StringIndexer
from pyspark.ml.feature import StandardScaler
from pyspark.ml import Pipeline

scaler = StandardScaler(inputCol="features", outputCol="scaled_features", withStd=True, withMean=True)
string_indexer = StringIndexer(inputCol="category", outputCol="index_category")
pipeline = Pipeline(stages=[string_indexer, scaler])

fitted_pipeline = pipeline.fit(raw_data) 
preprocessed_data = fitted_pipeline.transform(raw_data)
```
Once you've done this, you hold all preprocessed data in your data frame or RDD. While there's no proper slicing built in, afaik, you could ```zipWithIndex``` and filter on it, or even use SQL on the data frame to get what you want. See for instance the first comment in this one:
http://stackoverflow.com/questions/24677180/how-do-i-select-a-range-of-elements-in-spark-rdd

So, to wrap this up, I think this could be a viable choice. I'm not sure about performance, however, in terms of how long it takes to slice like this, but I can only assume it's a viable trade-off of speed for convenience.

**Comment by maxpumperla** [comment]
Right now I unfortunately find little time to follow recent advances in Keras, but I absolutely agree with you that something like this should exist without touch core. 

Anyway, I think we could quickly build such a data server API wrapping the above ideas in a module.

**Comment by EderSantana** [comment]
Thanks Max! 

Yeah, I got feathersjs dev to solve your issue :D check it out.

I'd be down to help with this data server API. I'll read your suggestions and see how I can help. Thanks for the quick reply!!

**Comment by EderSantana** [comment]
on that same subject, dask seems to be an opportunity as well https://github.com/tflearn/tflearn/blob/master/examples/basics/dask.py

**Comment by maxpumperla** [comment]
Yes, dask is very cool and probably much more straight forward.

**Comment by Humoyun** [comment]
Hello guys, 

I have some problems with elephas imports

when I imported :  from elephas.utils.rdd_utils import to_simple_rdd
I got:  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/elephas/utils/rdd_utils.py", line 3, in <module>
    from pyspark.mllib.regression import LabeledPoint
  File "/usr/local/spark/python/pyspark/mllib/__init__.py", line 27, in <module>
    raise Exception("MLlib requires NumPy 1.4+")
Exception: MLlib requires NumPy 1.4+

But my numpy is higher than 1.4+ (1.11), then I uninstalled numpy and reinstalled with pip (earlier numpy was installed with easy_install), this time pip installed numpy-1.6 and then "from elephas.utils.rdd_utils import to_simple_rdd" worked fine but this time I got another problem with "from keras.utils import np_utils"  it shows "module compiled against API version 7 but this version of numpy is 6". So, what version of numpy I should use for elephas works correctly. I really need your help

**Comment by maxpumperla** [comment]
Hi @Humoyun, that sounds really annoying. Generally, there shouldn't be such errors, which in particular makes them difficult to (remote) debug.

In the next few days I'll have to work on elephas a bit anyway, to bring it up to speed w.r.t. Keras 1.0. When doing that I'll try to reproduce what you have there or at least recommend a way to install things. Sounds OK?

**Comment by Humoyun** [comment]
@maxpumperla , I fixed this imports, but this time I encountered slice_X error, 
"ImportError: cannot import name slice_X"

**Comment by maxpumperla** [comment]
@Humoyun that is part of the keras 1.0 migration, slice_X resides in a different place now. I'll start with that soon.

**Comment by pucandrzej** [comment]
> 

Hi. For me the issue is that when I call train on the pipeline with elephas estimator & keras model the RAM disappears like crazy and 

Py4JError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe

happens. Any suggestions? Do You know what can be the issue?
